Cookbook/Proposals/ComposableReducers.md

```diff
@@ -0,0 +1,250 @@
+# Composable Reducers using Lenses and Prisms
+
+* Author: Yasuhiro Inami
+* Review Manager: David Rodrigues
+
+## Introduction
+
+This proposal adds functional [`Lens` and `Prism`](#functional-lens--prism) to break down `reducer` / `state` / `event` (`action`) in [ReactiveFeedback](https://github.com/Babylonpartners/ReactiveFeedback) so that we can create a large application from small component compositions.
+
+## Motivation
+
+In Babylon.app, we are using [ReactiveFeedback](https://github.com/Babylonpartners/ReactiveFeedback) to control states, side-effects, and feedback loops to define a particular screen (view controller) behavior, owned by `ViewModel`.
+
+However, this `ViewModel` can easily become too complex as `reducer: (State, Event) -> State` grows large with tons of pattern-matching.
+
+Unfortunately, splitting into multiple `ViewModel`s is not a clever solution, as managing multiple `ReactiveFeedback`s that interact with each other tends to be hard to control.
+
+To curcumvent this problem, we instead **split `reducer`s and also `state`s and `event`s**, and then combine them using `Lens` and `Prism`.
+
+The basic idea can be found in @mbrandonw ’s talk:
+
+[Brandon Williams \- Composable Reducers & Effects Systems \- YouTube](https://www.youtube.com/watch?v=QOIigosUNGU)
+
+## Proposed solution
+
+### Functional `Lens` & `Prism`
+
+```swift
+/// For accessing struct members.
+/// e.g. Whole = whole struct (members), Part = partial member
+struct Lens<Whole, Part> {
+    let get: (Whole) -> Part
+    let set: (Whole, Part) -> Whole
+}
+
+/// For accessing enum cases.
+/// e.g. Whole = all possible enum cases, B = partial case
+struct Prism<Whole, Part> {
```

Comment:
Are `preview` and `review` standard lingo for a `Prism`? I know that Brandon used these names in his presentation but I don’t think they describe the purpose of the two functions very well. To me, `subspace` and `update` reads more clearly.

Reply:
I think Brandon followed Haskell's naming for those: http://hackage.haskell.org/package/lens-4.17.1/docs/Control-Lens-Prism.html

Reply:
@zzcgumn 
Yes, `preview` and `review` are commonly used names in Haskell.

There are some alternatives for `Prism` method names which may be more intuitive:

1. Scala Monocle uses `getOption: Whole => Option[Part]` and `reverseGet (aka apply): Part => Whole`
    https://julien-truffaut.github.io/Monocle/optics/prism.html

2. `tryGet: Whole -> Part?` and `inject: Part -> Whole` is introduced here
    https://broomburgo.github.io/fun-ios/post/lenses-and-prisms-in-swift-a-pragmatic-approach/

I'm OK with renaming too (I personally prefer 2.), but we might want to stay as is, since there are many more upcoming layers in these functional Optics family as described in 
http://oleg.fi/gists/posts/2017-04-18-glassery.html

So renaming each of them may probably consume our brain too much that it's often better to just borrow the same names from original.